File: Python/Leetcode Medium/155MinStack.py
#https://leetcode.com/problems/min-stack/description/
class MinStack:

    def __init__(self):
        self.normal = []
        self.min = []

    def push(self, val: int) -> None:
        if len(self.normal) == 0:
            self.normal.append(val)
            self.min.append(val)
        else:
            self.normal.append(val)
            if val <= self.min[-1]:
                self.min.append(val) 
        
    def pop(self) -> None:
        val = self.normal.pop()
        if self.min[-1] == val:
            self.min.pop()
        return val

    def top(self) -> int:
        return self.normal[-1]

    def getMin(self) -> int:
        return self.min[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: C++/Leetcode Medium/155MinStack.cpp
#include <vector>
using namespace std;
class MinStack {
public:
    vector<int> stack;
    vector<int> minStack;
    MinStack() {
        
    }
    
    void push(int val) {
        stack.push_back(val);
        if (minStack.empty() || minStack.back() >= val) minStack.push_back(val);
    }
    
    void pop() {
        if (minStack.back() == stack.back()) minStack.pop_back();
        stack.pop_back();
    }
    
    int top() {
        return stack.back();
    }
    
    int getMin() {
        return minStack.back();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

#include <vector>
#include <algorithm>
using namespace std;

class MinStack {
public:
    vector<int> stack;
    vector<int> minStack;
    MinStack() {

    }
    
    void push(int val) {
        stack.push_back(val);
        if (minStack.empty()) minStack.push_back(val);
        else {
            if (val <= minStack.back()) minStack.push_back(val);
        }
    }
    
    void pop() {
        if (stack.back() == minStack.back()) minStack.pop_back();
        stack.pop_back();
    }
    
    int top() {
        return stack.back();
    }
    
    int getMin() {
        return minStack.back();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

#include <vector>
#include <tuple>
using namespace std;
class MinStack {
public:

    vector<tuple<int, int>> stack;
    //1st is num, 2nd is minNum
    MinStack() {}
    
    void push(int val) {
        if (stack.size() == 0 || get<1>(stack.back()) > val) stack.emplace_back(val, val);
        else stack.emplace_back(val, get<1>(stack.back()));
    }
    
    void pop() {
        stack.pop_back();
    }
    
    int top() {
        return get<0>(stack.back());
    }
    
    int getMin() {
        return get<1>(stack.back());
    }
};